File: web/lib/voice/transcript-gate.ts
/** Gate STT finals before sending to brain — reduces echo fragments and junk turns. */

import { wordCount } from "@/lib/early-tts";

const INDIC_SCRIPT = /[\u0900-\u097F\u0C00-\u0C7F]/;

/** Count words; Indic scripts often lack spaces — use char clusters as fallback. */
export function effectiveWordCount(text: string): number {
  const trimmed = text.trim();
  if (!trimmed) return 0;
  const spaced = wordCount(trimmed);
  if (spaced >= 2) return spaced;
  if (INDIC_SCRIPT.test(trimmed)) {
    const clusters = trimmed.replace(/\s+/g, "").length;
    if (clusters >= 6) return Math.max(2, Math.ceil(clusters / 4));
  }
  return spaced;
}

export type TranscriptGateOpts = {
  /** After explicit barge-in — allow shorter caller utterances. */
  afterBarge?: boolean;
  /** From queue during agent TTS — require more substance. */
  fromSpeechQueue?: boolean;
};

/**
 * True when a final transcript is worth a brain turn.
 * Blocks STT debris like "and that will plot." right after agent speech.
 */
export function isSubstantiveTranscript(text: string, opts?: TranscriptGateOpts): boolean {
  const t = text.trim();
  if (!t) return false;
  const words = effectiveWordCount(t);
  const chars = t.replace(/\s+/g, "").length;

  if (opts?.afterBarge) {
    return words >= 1 && chars >= 3;
  }
  if (opts?.fromSpeechQueue) {
    return words >= 3 || chars >= 12;
  }
  return words >= 2 || chars >= 8;
}

const VALID_SHORT_REPLY =
  /^(?:ok|okay|yes|no|sure|thanks|thank you|got it|sare|hmm|mm|aha|అవును|సరే|ठीक|हाँ)\.?$/i;
const MIDWORD_STUMP = /\b(?:lak|cro|rupee|from|with|and|or|the|we|i|to|for|a|an)\.?$/i;

/** Very short assistant text that looks cut off mid-thought. */
export function looksIncompleteReply(text: string): boolean {
  const t = text.trim();
  if (!t) return false;
  if (VALID_SHORT_REPLY.test(t)) return false;
  // Token/barge crumbs that end with forced "." still count as incomplete.
  if (MIDWORD_STUMP.test(t) && t.length < 100) {
    if (t.length < 24 || /\b(?:lak|cro)\.?$/i.test(t)) return true;
  }
  if (t.length >= 40) return false;
  if (/[.!?।]\s*$/.test(t)) return false;
  if (effectiveWordCount(t) >= 8) return false;
  return true;
}

File: server/services/transcript_gate.py
"""Gate STT finals before launching a brain turn — mirrors web transcript-gate.ts."""
from __future__ import annotations

import re

# Devanagari + Telugu (same ranges as web).
_INDIC_SCRIPT = re.compile(r"[\u0900-\u097F\u0C00-\u0C7F]")
_WS = re.compile(r"\s+")
_SHORT_RESPONSE = re.compile(
    r"^(?:yes|no|yeah|yep|nope|wait|stop|కాదు|లేదు|లేదండి|అవును|వద్దు|नहीं|हाँ|रुको|ठीक)[.!?,]*$",
    re.I,
)


def effective_word_count(text: str) -> int:
    """Count words; Indic scripts often lack spaces — use char clusters as fallback."""
    trimmed = (text or "").strip()
    if not trimmed:
        return 0
    spaced = len([w for w in _WS.split(trimmed) if w])
    if spaced >= 2:
        return spaced
    if _INDIC_SCRIPT.search(trimmed):
        clusters = len(_WS.sub("", trimmed))
        if clusters >= 6:
            return max(2, (clusters + 3) // 4)
    return spaced


def is_substantive_transcript(
    text: str,
    *,
    after_barge: bool = False,
    from_speech_queue: bool = False,
) -> bool:
    """True when a final is worth a brain turn (blocks junk / echo fragments)."""
    t = (text or "").strip()
    if not t:
        return False
    if _SHORT_RESPONSE.fullmatch(t):
        return True
    words = effective_word_count(t)
    chars = len(_WS.sub("", t))
    if after_barge:
        return words >= 1 and chars >= 3
    if from_speech_queue:
        return words >= 3 or chars >= 12
    return words >= 2 or chars >= 8
